google/cloud/internal/rest_completion_queue_impl.cc

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/rest_completion_queue_impl.h"
+#include "google/cloud/internal/timer_queue.h"
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+RestCompletionQueueImpl::RestCompletionQueueImpl()
+    : tq_(std::make_shared<internal::TimerQueue>()) {}
+
+void RestCompletionQueueImpl::Run() { tq_->Service(); }
+
+void RestCompletionQueueImpl::Shutdown() { tq_->Shutdown(); }
+
+void RestCompletionQueueImpl::CancelAll() { tq_->CancelAll(); }
+
+future<StatusOr<std::chrono::system_clock::time_point>>
+RestCompletionQueueImpl::MakeDeadlineTimer(
+    std::chrono::system_clock::time_point deadline) {
+  return tq_->Schedule(deadline);
+}
+
+future<StatusOr<std::chrono::system_clock::time_point>>
+RestCompletionQueueImpl::MakeRelativeTimer(std::chrono::nanoseconds duration) {
+  using std::chrono::system_clock;
+  auto const d = std::chrono::duration_cast<system_clock::duration>(duration);
```

Comment:
This should probably be `std::chrono::ceil<system_clock::duration>(duration)`, so timers don't fire early, but that is C++17 talk.

In the meantime, something like ...
```
  if (d < duration) d += system_clock::duration{1};
```
should suffice.